vulnerabilities/import_runner.py

```diff
@@ -0,0 +1,211 @@
+
+        NB: Data sources provide two kinds of records; vulnerabilities and packages. Vulnerabilities are potentially
+        shared across many packages, from the same data source and from different data sources. For example, a
+        vulnerability in the Linux kernel is mentioned by advisories from all Linux distributions that package this
+        kernel version.
+        """
+        logger.debug(f'Starting import for {self.importer.name}.')
+        data_source = self.importer.make_data_source(cutoff_date=cutoff_date, batch_size=self.batch_size)
+
+        with data_source as ds:
+            for batch in ds.new_records():
+                impacted, resolved = _collect_packages_from_batch(batch)
+                impacted, resolved = _bulk_insert_packages(impacted, resolved)
+
+                vulnerabilities = _insert_vulnerabilities_and_references(batch)
+
+                _bulk_insert_impacted_and_resolved_packages(batch, vulnerabilities, impacted, resolved)
+
+            self.importer.last_run = datetime.datetime.utcnow()
+            self.importer.save()
+
+        logger.debug(f'Successfully finished import for {self.importer.name}.')
+
+
+def _bulk_insert_impacted_and_resolved_packages(
+    batch: Sequence[VulnerabilityInfo],
+    vulnerability_models: Set[models.Vulnerability],
+    impacted_package_models: Mapping[str, models.Package],
+    resolved_package_models: Mapping[str, models.Package],
+) -> None:
+
+    impacted_refs: List[models.ImpactedPackage] = []
+    resolved_refs: List[models.ResolvedPackage] = []
+
+    for vuln_info in batch:
+        vuln_model = _vuln_info_to_vuln_model(vuln_info, vulnerability_models)
+        vulnerability_models.remove(vuln_model)  # minor optimization
+
+        for impacted_package in vuln_info.impacted_packages:
+            ip = models.ImpactedPackage(
+                vulnerability=vuln_model,
+                package=impacted_package_models[impacted_package.package_url]
+            )
+            impacted_refs.append(ip)
+
+        for resolved_package in vuln_info.resolved_packages:
+            ip = models.ResolvedPackage(
+                vulnerability=vuln_model,
+                package=resolved_package_models[resolved_package.package_url]
+            )
+            resolved_refs.append(ip)
+
+    models.ImpactedPackage.objects.bulk_create(impacted_refs)
+    models.ResolvedPackage.objects.bulk_create(resolved_refs)
+
+
+@transaction.atomic
+def _insert_vulnerabilities_and_references(batch: Sequence[VulnerabilityInfo]) -> Set[models.Vulnerability]:
+    """
+    TODO Consider refactoring to use bulk_create() and avoid get_or_create() when possible.
+    """
+    vulnerabilities = set()
+
+    for vuln_info in batch:
+        vuln: models.Vulnerability
+
+        if vuln_info.cve_id:
+            vuln, created = models.Vulnerability.objects.get_or_create(cve_id=vuln_info.cve_id)
+            if created:
+                vuln.summary = vuln_info.summary
```

Comment:
We need some way  to determine, which `summary` is better and then decide to update the summary or leave it as it is.

Reply:
I'm not optimistic there is one. But even if there was a way, if all we have is the summary, we can't detect that a new vulnerability is actually an update to an existing vulnerability.

Reply:
I was thinking of moving `summary` from `Vulnerability` model to the `ImpactedPackage` model . So the context won't be lost.

Reply:
In theory it makes sense that the same vulnerability could be described in different terms, depending on the affected package. But in practice, do any of the source we ingest so far do this? Or would we end up copying the same description to every `ImpactedPackage`? That wouldn't really be an issue though.

Fundamentally the questions are:
* What is the definition of "vulnerability"?
* What is the minimum number of attributes (and what are those) for identifying a vulnerability?